File: backend/api/views/transactions.py
"""List, create, update and delete the signed-in user's transactions."""

from datetime import date

from django.db.models import Q
from rest_framework import status
from rest_framework.exceptions import ValidationError
from rest_framework.pagination import LimitOffsetPagination
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response
from rest_framework.views import APIView

from ..models import Transaction
from ..serializers import TransactionSerializer

# Upper bound for one page of the transactions list; the SPA asks for 10.
TRANSACTIONS_MAX_LIMIT = 100


class OptionalLimitOffsetPagination(LimitOffsetPagination):
    """Paginate only when the client asks for it.

    With ``default_limit`` unset, ``paginate_queryset`` returns None unless a
    ``?limit=`` param is present, so the dashboard keeps receiving the plain
    array it always did while the transactions page opts into
    ``{count, next, previous, results}`` pages.
    """

    default_limit = None
    max_limit = TRANSACTIONS_MAX_LIMIT


def _parse_iso_date(params, key):
    """Return ``params[key]`` as a date, None if absent, or raise a 400."""
    raw = params.get(key)
    if not raw:
        return None
    try:
        return date.fromisoformat(raw)
    except ValueError:
        raise ValidationError({key: 'Use the format YYYY-MM-DD.'})


class TransactionView(APIView):
    permission_classes = [IsAuthenticated]
    pagination_class = OptionalLimitOffsetPagination

    def filter_queryset(self, request):
        """Apply the optional query-string filters of the transactions page.

        ``category`` / ``type`` match exactly, ``search`` is a case-insensitive
        substring match on title or notes, ``date_from`` / ``date_to`` are
        inclusive ISO dates. Unknown category/type values simply match nothing.
        """
        params = request.query_params
        # Explicit secondary key so limit/offset pages never overlap or skip
        # rows that share the same date.
        queryset = request.user.transactions.order_by('-date', '-id')

        if category := params.get('category'):
            queryset = queryset.filter(category=category)

        if tx_type := params.get('type'):
            queryset = queryset.filter(type=tx_type)

        if search := params.get('search', '').strip():
            queryset = queryset.filter(Q(title__icontains=search) | Q(notes__icontains=search))

        if date_from := _parse_iso_date(params, 'date_from'):
            queryset = queryset.filter(date__gte=date_from)

        if date_to := _parse_iso_date(params, 'date_to'):
            queryset = queryset.filter(date__lte=date_to)

        return queryset

    def get(self, request):
        transactions = self.filter_queryset(request)

        paginator = self.pagination_class()
        page = paginator.paginate_queryset(transactions, request, view=self)
        if page is not None:
            serializer = TransactionSerializer(page, many=True)
            return paginator.get_paginated_response(serializer.data)

        serializer = TransactionSerializer(transactions, many=True)
        return Response(serializer.data, status=status.HTTP_200_OK)

    def post(self, request):
        serializer = TransactionSerializer(data=request.data, context={'request': request})
        if serializer.is_valid():
            serializer.save()
            return Response(serializer.data, status=status.HTTP_201_CREATED)
        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)


class TransactionDetailView(APIView):
    permission_classes = [IsAuthenticated]

    def get_object(self, request, pk):
        try:
            return request.user.transactions.get(pk=pk)
        except Transaction.DoesNotExist:
            return None

    def put(self, request, pk):
        transaction = self.get_object(request, pk)
        if transaction is None:
            return Response(status=status.HTTP_404_NOT_FOUND)
        serializer = TransactionSerializer(
            transaction, data=request.data, context={'request': request}
        )
        if serializer.is_valid():
            serializer.save()
            return Response(serializer.data, status=status.HTTP_200_OK)
        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)

    def delete(self, request, pk):
        transaction = self.get_object(request, pk)
        if transaction is None:
            return Response(status=status.HTTP_404_NOT_FOUND)
        transaction.delete()
        return Response(status=status.HTTP_204_NO_CONTENT)

File: app-frontend/src/lib/api/transactions.ts
import apiClient from '@/lib/apiClient'
import type { MonthlyData, PaginatedResponse, Transaction } from '@/lib/types'
import { monthBounds } from '@/lib/utils'

/** Query-string filters understood by `GET /api/transactions/`. */
export interface TransactionQuery {
  search?: string
  category?: string
  type?: Transaction['type']
  /** Inclusive ISO dates. */
  date_from?: string
  date_to?: string
}

export type NewTransaction = Omit<Transaction, 'id'>

/** Plain array; used when the caller wants every row matching the query. */
export async function fetchTransactions(query: TransactionQuery = {}): Promise<Transaction[]> {
  const response = await apiClient.get<Transaction[]>('/transactions/', { params: query })
  return response.data
}

/** One limit/offset page; the server only wraps the response when `limit` is set. */
export async function fetchTransactionPage(
  query: TransactionQuery,
  limit: number,
  offset: number,
): Promise<PaginatedResponse<Transaction>> {
  const response = await apiClient.get<PaginatedResponse<Transaction>>('/transactions/', {
    params: { ...query, limit, offset },
  })
  return response.data
}

/** Every transaction of the month containing `month`. */
export function fetchMonthTransactions(month: Date): Promise<Transaction[]> {
  const { from, to } = monthBounds(month)
  return fetchTransactions({ date_from: from, date_to: to })
}

/** The newest `count` transactions across all months. */
export async function fetchRecentTransactions(count: number): Promise<Transaction[]> {
  const page = await fetchTransactionPage({}, count, 0)
  return page.results
}

export async function createTransaction(transaction: NewTransaction): Promise<Transaction> {
  const response = await apiClient.post<Transaction>('/transactions/', transaction)
  return response.data
}

export async function updateTransaction(transaction: Transaction): Promise<Transaction> {
  const response = await apiClient.put<Transaction>(`/transactions/${transaction.id}/`, transaction)
  return response.data
}

export async function deleteTransaction(id: string): Promise<void> {
  await apiClient.delete(`/transactions/${id}/`)
}

export async function fetchMonthlySummary(): Promise<MonthlyData[]> {
  const response = await apiClient.get<MonthlyData[]>('/monthly-summary/')
  return response.data
}
